x/gamm/keeper/msg_server_test.go

```diff
@@ -356,3 +362,211 @@ func (suite *KeeperTestSuite) TestExitPool_Events() {
+		clPoolUsdcBalanceBeforeMigration := suite.App.BankKeeper.GetBalance(suite.Ctx, clPoolAddress, USDC)
+
+		// Reset event counts to 0 by creating a new manager.
+		suite.Ctx = suite.Ctx.WithEventManager(sdk.NewEventManager())
+		suite.Require().Equal(0, len(suite.Ctx.EventManager().Events()))
+
+		// Migrate the user's gamm shares to a full range concentrated liquidity position
+		resp, err := msgServer.MigrateSharesToFullRangeConcentratedPosition(sdk.WrapSDKContext(suite.Ctx), msg)
+		if test.expectedErr != nil {
+			suite.Require().Error(err)
+			suite.Require().ErrorContains(err, test.expectedErr.Error())
+			continue
+		}
+
+		suite.Require().NoError(err)
+
+		// Assert events are emitted
+		suite.AssertEventEmitted(suite.Ctx, types.TypeEvtMigrateShares, test.expectedMigrateShareEvents)
+		suite.AssertEventEmitted(suite.Ctx, sdk.EventTypeMessage, test.expectedMessageEvents)
+
+		// Note gamm pool balance after migration
+		gammPoolEthBalancePostMigrate := suite.App.BankKeeper.GetBalance(suite.Ctx, gammPoolAddress, ETH)
+		gammPoolUsdcBalancePostMigrate := suite.App.BankKeeper.GetBalance(suite.Ctx, gammPoolAddress, USDC)
+
+		// Note user balance after migration
+		userEthBalanceTransferredToClPool := gammPoolEthBalancePostJoin.Sub(gammPoolEthBalancePostMigrate)
+		userUsdcBalanceTransferredToClPool := gammPoolUsdcBalancePostJoin.Sub(gammPoolUsdcBalancePostMigrate)
+
+		// Note cl pool balance after migration
+		clPoolEthBalanceAfterMigration := suite.App.BankKeeper.GetBalance(suite.Ctx, clPoolAddress, ETH)
+		clPoolUsdcBalanceAfterMigration := suite.App.BankKeeper.GetBalance(suite.Ctx, clPoolAddress, USDC)
+
+		// The balance in the cl pool should be equal to what the user previously had in the gamm pool
+		poolEthBalanceInTolerance := userEthBalanceTransferredToClPool.Amount.Sub(clPoolEthBalanceAfterMigration.Amount).LTE(sdk.NewInt(1))
+		poolUsdcBalanceInTolerance := userEthBalanceTransferredToClPool.Amount.Sub(clPoolEthBalanceAfterMigration.Amount).LTE(sdk.NewInt(1))
+		fmt.Printf("userEthBalanceTransferredToClPool.Amount %v \n", userEthBalanceTransferredToClPool.Amount)
+		fmt.Printf("clPoolEthBalanceAfterMigration.Amount %v \n", clPoolEthBalanceAfterMigration.Amount)
+		fmt.Printf("userUsdcBalanceTransferredToClPool.Amount %v \n", userUsdcBalanceTransferredToClPool.Amount)
+		fmt.Printf("clPoolUsdcBalanceAfterMigration.Amount %v \n", clPoolUsdcBalanceAfterMigration.Amount)
+		fmt.Printf("poolUsdcBalanceInTolerance %v \n", poolUsdcBalanceInTolerance)
+		suite.Require().True(poolEthBalanceInTolerance)
+		suite.Require().True(poolUsdcBalanceInTolerance)
+		// suite.Require().Equal(userEthBalanceTransferredToClPool.String(), clPoolEthBalanceAfterMigration.String())
+		// suite.Require().Equal(userUsdcBalanceTransferredToClPool.String(), clPoolUsdcBalanceAfterMigration.String())
+
+		// Withdraw the user's concentrated liquidity position
+		minTick, maxTick := cl.GetMinAndMaxTicksFromExponentAtPriceOne(clPool.GetPrecisionFactorAtPriceOne())
```

Comment:
IMO doing all these checks without withdrawing feels like better unit testing process, as we're adding another secondary logic gate for the checks, instead of uniquely testing the SUT. It mught be because of a reason I do not know of or just a difference in perspective so feel free to ignore if you have different view points

Reply:
I would agree with this, but I don't have a simple way right now of checking what cl liquidity corresponds to in terms of they underlying. Because of this, I need to withdraw to determine how much of the underlying the user is allotted.

Reply:
After thinking about this for a little, I agree this test overstretched for no good reason. I will reduce it to what you suggested, thank you!

Reply:
https://github.com/osmosis-labs/osmosis/pull/4051/commits/de1590c38fb8c5f801f5528677464682a79840e3